sys/linux/smack.txt

```diff
@@ -1,21 +1,213 @@
+#type smack_perm stringnoz
+type smack_perm stringnoz[perm]
+
+smackfs_access {
+	subject	smack_labelnoz
+	sp0	const[' ', int8]
+	object	smack_labelnoz
+	sp1	const[' ', int8]
+	access	smack_perm
+	z	const[0, int8]
+} [packed]
+
+smackfs_change_access {
+	subject	smack_labelnoz
+	sp0	const[' ', int8]
+	object	smack_labelnoz
+	sp1	const[' ', int8]
+	allow	smack_perm
+	sp2	const[' ', int8]
+	deny	smack_perm
+	z	const[0, int8]
+} [packed]
+
+resource fd_smackfs_access[fd]
+smackfs_check_access = "/sys/fs/smackfs/access", "/sys/fs/smackfs/access2"
+openat$smackfs_access(fd const[AT_FDCWD], file ptr[in, string[smackfs_check_access]], flags const[O_RDWR], mode const[0]) fd_smackfs_access
+write$smackfs_access(fd fd_smackfs_access, buf ptr[in, smackfs_access], count len[buf])
+read$smackfs_access(fd fd_smackfs_access, buf ptr[out, fmt[dec, int32[0:1]]], count len[buf])
+
+# load, load2, load-self, load-self2
+resource fd_smackfs_load[fd]
+smackfs_load_files = "/sys/fs/smackfs/load", "/sys/fs/smackfs/load2", "/sys/fs/smackfs/load-self", "/sys/fs/smackfs/load-self2"
+openat$smackfs_load(fd const[AT_FDCWD], file ptr[in, string[smackfs_load_files]], flags const[O_RDWR], mode const[0]) fd_smackfs_load
+write$smackfs_load(fd fd_smackfs_load, buf ptr[in, smackfs_access], count len[buf])
+
+# ambient, unconfined, syslog, revoke-subject
+resource fd_smackfs_label[fd]
+smackfs_label_files = "/sys/fs/smackfs/ambient", "/sys/fs/smackfs/syslog", "/sys/fs/smackfs/unconfined", "/sys/fs/smackfs/revoke-subject"
+openat$smackfs_label(fd const[AT_FDCWD], file ptr[in, string[smackfs_label_files]], flags const[O_RDWR], mode const[0]) fd_smackfs_label
+write$smackfs_label(fd fd_smackfs_label, buf ptr[in, smack_label], count len[buf])
+
+resource fd_smackfs_change_rule[fd]
+openat$smackfs_change_rule(fd const[AT_FDCWD], file ptr[in, string["/sys/fs/smackfs/change-rule"]], flags const[O_RDWR], mode const[0]) fd_smackfs_change_rule
+write$smackfs_change_rule(fd fd_smackfs_change_rule, buf ptr[in, smackfs_change_access], count len[buf])
+
+smackfs_cipso_category {
+	cat	fmt[dec, int8[0:SMACK_CIPSO_MAXCATNUM]]
+	sp	const[' ', int8]
+}
+
+smackfs_cipso {
+	label	smack_labelnoz
+	sp0	const[' ', int8]
+# SMK_DIGITLEN == 4
+	level	fmt[dec, int8[0:SMACK_CIPSO_MAXLEVEL]]
+	sp1	const[' ', int8]
+# SMK_DIGITLEN
+# TODO: num	fmt[dec, len[cats, int8]]
```

Comment:
Why doesn't this work? It should, no?

Reply:
It doesn't work:
```
$ make generate
smack.txt:88:15: wrong number of arguments for type len, expect len target
```
This works:
```
num	fmt[dec, len[cats]]
```
I think that it will use int32 or intptr in this case.

However, we need int8 here because of: https://www.kernel.org/doc/html/latest/admin-guide/LSM/Smack.html
> cipso2
> This interface allows a specific CIPSO header to be assigned to a Smack label. The format accepted on write is:
> 
> "%s%4d%4d"["%4d"]...

Actually, it's not a "%4d" it's "%3d" with whitespace. This len is in [0:184] https://elixir.bootlin.com/linux/v5.4-rc8/source/security/smack/smackfs.c#L887

Reply:
Using `fmt[dec, len[cats]]` looks better to me. We will have the correct length and unless the array itself contains more than 184 elements, this value will be correct as well.

Reply:
Done. I just hope it will not add leading zeroes. It's not clear to me what input will be generated in this case.

> "fmt": a string representation of an integer (not zero-terminated), type-options:
> 	format (one of "dec", "hex", "oct") and the value (a resource, int, flags, const or proc)
> 	the resulting data is always fixed-size (formatted as "%020llu", "0x%016llx" or "%023llo", respectively)

Smack expects here exactly 3-size number, e.g.: 001, 183, 010...

Reply:
Ah, I see. Yes, it won't format numbers this ways, but int8 won't help either.
The problem is that we still don't support for dynamically-sized arguments. Say an array is variable size, but once a program is generated its size is known. Whereas if we want to sprintf a dynamic number (e.g. a pid) we don't know until the runtime how many digits it will take. So the current format is a simplified version that always pads numbers to max size.
But actually this case is yet different b/c for %03d we do know the resulting size.

Reply:
I've clarified the comment a bit:
https://github.com/google/syzkaller/commit/0199dc9d8c15c7abbffdb9644cbebd2acb6f8a67